### PR TITLE
WIP: Rewrite to use child process

### PR DESCRIPTION
**Pull request by steelbrain**
Instead of doing everything in the renderer process, we're not linting in a separate worker process.

Pinging so they would be subscribed, @Arcanemagus @SpainTrain

[Review on Reviewable](https://reviewable.io/reviews/atomlinter/linter-eslint/258)

### CODE REVIEW

**Comment by SpainTrain** [comment]
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[.gitignore, line 3 \[r1\]](https://reviewable.io:443/reviews/atomlinter/linter-eslint/258#-K2sANCEfOhJZbqIkH46)** ([raw file](https://github.com/atomlinter/linter-eslint/blob/59bc20c3c10e23fd84bd4a8c20fdd83e9fedca76/.gitignore#L3)):</sup>
nit: IMO IDE-specific ignores should go in one's home directory gitignore since it is related to the environment, not the project


---

<sup>**[package.json, line 3 \[r1\]](https://reviewable.io:443/reviews/atomlinter/linter-eslint/258#-K2sB5Z9mxPXWi7U469L)** ([raw file](https://github.com/atomlinter/linter-eslint/blob/59bc20c3c10e23fd84bd4a8c20fdd83e9fedca76/package.json#L3)):</sup>
:+1: 


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/atomlinter/linter-eslint/258)

**Comment by steelbrain** [comment]
So the process spawning and inter-process communication is working great, now we just have to hook the `eslint` wires in the pure node process

**Comment by steelbrain** [comment]
Review status: 1 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[.gitignore, line 3 \[r1\]](https://reviewable.io:443/reviews/atomlinter/linter-eslint/258#-K2sANCEfOhJZbqIkH46)** ([raw file](https://github.com/atomlinter/linter-eslint/blob/59bc20c3c10e23fd84bd4a8c20fdd83e9fedca76/.gitignore#L3)):</sup>
I do agree, I'll be careful about it next time


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/atomlinter/linter-eslint/258)

**Comment by steelbrain** [comment]
The worker and eslint reporting is working now, all we need to do is re-implement the config from last time and parse the errors we get from eslint

**Comment by steelbrain** [comment]
Blocked at https://github.com/eslint/eslint/issues/4394

**Comment by steelbrain** [comment]
ToDo
  - [ ] Delay spawning of the worker thread by a few seconds
  - [ ] Implement rule id in messages
  - [ ] Remove old file

after that, it'll be good to merge

**Comment by SpainTrain** [comment]
Reviewed 2 of 4 files at r3, 2 of 3 files at r4, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/atomlinter/linter-eslint/258)

**Comment by steelbrain** [comment]
Alright, I just tested it and it works for me, r?

**Comment by Arcanemagus** [comment]
Looks like you have a missing dependency.

```
Error: Cannot find module 'childprocess-promise'
    at Module._resolveFilename (module.js:334:15)
    at Function.Module._resolveFilename (C:\Users\Landon Abney\AppData\Local\atom\app-1.1.0\resources\app.asar\src\module-cache.js:383:52)
    at Function.Module._load (module.js:284:25)
    at Module.require (module.js:363:17)
    at require (module.js:382:17)
    at Object.<anonymous> (C:/Users/Landon Abney/Documents/GitHub/linter-eslint/lib/helpers.js:3:41)
    at Module._compile (module.js:428:26)
    at Object.defineProperty.value [as .js] (C:\Users\Landon Abney\AppData\Local\atom\app-1.1.0\resources\app.asar\src\compile-cache.js:169:21)
    at Module.load (module.js:353:32)
    at Function.Module._load (module.js:308:12)
    at Module.require (module.js:363:17)
    at require (module.js:382:17)
    at Object.<anonymous> (C:/Users/Landon Abney/Documents/GitHub/linter-eslint/lib/main.js:4:41)
    at Module._compile (module.js:428:26)
    at Object.defineProperty.value [as .js] (C:\Users\Landon Abney\AppData\Local\atom\app-1.1.0\resources\app.asar\src\compile-cache.js:169:21)
    at Module.load (module.js:353:32)
```

**Comment by Arcanemagus** [comment]
You should remove the `coffeelint.json` and put in a `.eslintrc` as there is no longer any CoffeeScript code in this package after this PR.

**Comment by Arcanemagus** [comment]
The README also needs updating to match the state of this PR:
* `linter-eslint` isn't escaped for one of the instances in the Installation section.
* `eslintRulesDir` is removed from the configuration.
* `globalNodePath` is removed from the configuration.
* `lintHtmlFiles` has been added
* (Or just remove that entire settings section as most users will use the UI, and those that don't should already know what they are doing)
* Remove mention of `coffeelint`

File: lib/main.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import Path from 'path'
+import {CompositeDisposable} from 'atom'
+import {spawnWorker} from './helpers'
+import escapeHTML from 'escape-html'
+
+export default {
+  config: {
+    lintHtmlFiles: {
+      title: 'Lint HTML Files',
+      description: 'You should also add `eslint-plugin-html` to you .eslintrc plugins',
```
Comment (Arcanemagus) [inline lib/main.js]: you -> your

Comment (steelbrain) [inline lib/main.js]: Fixed thanks!

File: package.json [hunk 2]
```diff
@@ -10,8 +10,10 @@
     "atom": ">0.50.0"
   },
   "dependencies": {
-    "atom-linter": "^3.3.0",
-    "atom-package-deps": "^2.1.3",
+    "atom-linter": "^3.3.2",
+    "atom-package-deps": "^3.0.1",
+    "childprocess-promise": "^3.0.0",
+    "escape-html": "^1.0.3",
     "eslint": "^1.7.3",
```
Comment (Arcanemagus) [inline package.json]: Why not update the dependency while we are at it?

Comment (steelbrain) [inline package.json]: Sure

File: lib/main.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      default: true
+    },
+    disableWhenNoEslintrcFileInPath: {
+      title: 'Disable when no .eslintrc is found',
+      type: 'boolean',
+      default: true
+    }
+  },
+  activate: function() {
+    require('atom-package-deps').install('linter-eslint')
```
Comment (Arcanemagus) [inline lib/main.js]: No need to specify the name here with the current version of `atom-package-deps`.

Comment (steelbrain) [inline lib/main.js]: Done, Thanks :bow:

File: lib/main.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      if (lintHtmlFiles) {
+        this.scopes.push(embeddedScope)
+      } else {
+        if (this.scopes.indexOf(embeddedScope) !== -1) {
+          this.scopes.splice(this.scopes.indexOf(embeddedScope), 1)
+        }
+      }
+    }))
+
+    // Reason: I (steelbrain) have observed that if we spawn a process while atom is starting up, it can increase startup time by several seconds,
```
Comment (Arcanemagus) [inline lib/main.js]: Mind splitting this across lines a bit earlier? If I have to scroll to see it on GitHub it's too long :stuck_out_tongue:.

**Comment by steelbrain** [comment]
so we have two ways ahead of us, either append the rule id to message like `something bad happened (rule-bad)` and have it text like, you can see it [here](https://cloud.githubusercontent.com/assets/4278113/11153745/2bbe8946-89f8-11e5-820c-baef57da022a.png) or we can add a new dependency, use html and add a badge of rule id and escape the message like [this](https://cloud.githubusercontent.com/assets/4278113/11153765/5dea5918-89f8-11e5-8352-ee1eaf86a9fe.png)

File: package.json [hunk 5]
```diff
@@ -10,9 +10,11 @@
...
   "dependencies": {
-    "atom-linter": "^3.3.0",
-    "atom-package-deps": "^2.1.3",
-    "eslint": "^1.7.3",
+    "atom-linter": "^3.3.2",
+    "atom-package-deps": "^3.0.1",
+    "childprocess-promise": "^3.0.0",
+    "escape-html": "^1.0.3",
+    "eslint": "^1.9.0",
     "eslint-plugin-html": "^1.0.4",
```
Comment (Arcanemagus) [inline package.json]: This has a v1.1.0 out.

File: package.json [hunk 6]
```diff
@@ -10,9 +10,11 @@
...
     "loophole": "^1.1.0",
     "resolve": "^1.1.5"
```
Comment (Arcanemagus) [inline package.json]: v1.1.6 is out.

File: package.json (hunk 5 above)
Comment (steelbrain) [inline package.json]: I am about to remove this one from the package

**Comment by Arcanemagus** [comment]
I would much prefer the second method as it looks very nice, but it's not critical.

File: lib/worker.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      eslint = require(Path.join(eslintPath, 'node_modules', 'eslint', 'lib', 'cli.js'))
+      oldEslintPath = eslintPath
+    } catch (e) {
+      throw new Error('ESLint not found, Please install or make sure Atom is getting $PATH correctly')
+    }
+    oldGlobal = params.global
+  }
+
+  job.Response = new Promise(function(resolve) {
+    process.argv = [process.execPath, oldEslintPath, '--stdin', '--format', Path.join(__dirname, 'reporter.js'), params.filePath]
```
Comment (Arcanemagus) [inline lib/worker.js]: Mind splitting this up onto multiple lines?

**Comment by Arcanemagus** [comment]
LGTM after the above comment is resolved.

File: lib/worker.js (hunk 7 above)
Comment (steelbrain) [inline lib/worker.js]: done

Comment (Arcanemagus) [inline lib/worker.js]: Although the one you changed was a long one, this is the line that is (still) causing a scroll bar on GitHub.

Comment (steelbrain) [inline lib/worker.js]: Damn :stuck_out_tongue:

Comment (steelbrain) [inline lib/worker.js]: Done

**Comment by Arcanemagus** [comment]
Still left TODO that I see:
- [ ] Update the README.md
- [ ] Add an `.eslintrc` configuration

File: lib/main.js [hunk 8]
```diff
@@ -0,0 +1,114 @@
+'use babel'
+
+/* global atom */
```
Comment (Arcanemagus) [inline lib/main.js]: Shouldn't this just go in the eslint config?

File: .eslintrc [hunk 9]
```diff
@@ -0,0 +1,3 @@
+{
+  extends: "steelbrain"
```
Comment (Arcanemagus) [inline .eslintrc]: My `eslint` has no idea what you are talking about :stuck_out_tongue:.

Comment (steelbrain) [inline .eslintrc]:
```
npm install -g eslint-config-steelbrain
```
then use the global eslint from settings and then lint :stuck_out_tongue:

Comment (Arcanemagus) [inline .eslintrc]: Then you should specify that as a dev dependency, as it stands this `.eslintrc` means nothing...

Comment (steelbrain) [inline .eslintrc]: done

Comment (steelbrain) [inline .eslintrc]: by the way, if it's not installer and you try to lint with eslint, it cries out loud that `eslint-config-steelbrain` is missing :wink:

**Comment by steelbrain** [comment]
Merging after confirmed reports of it working on Windows and Ubuntu

**Pull request merged**